File: core/src/main/java/com/toocol/termio/core/mosh/core/statesnyc/UserStream.java
package com.toocol.termio.core.mosh.core.statesnyc;

import com.google.protobuf.ByteString;
import com.toocol.termio.core.mosh.core.proto.UserInputPB;

import java.util.Deque;
import java.util.Iterator;
import java.util.concurrent.ConcurrentLinkedDeque;

/**
 * user.h:UserStream
 *
 * @author ：JoeZane (joezane.cn@gmail.com)
 * @date: 2022/5/11 0:30
 * @version: 0.0.1
 */
public final class UserStream extends State<UserStream> {

    private final Deque<UserEvent> actions = new ConcurrentLinkedDeque<>();

    @Override
    public void subtract(UserStream prefix) {
        if (this.equals(prefix)) {
            actions.clear();
            return;
        }

        for (UserEvent next : prefix.actions) {
            UserEvent peek = actions.peek();
            if (peek != null && peek.equals(next)) {
                actions.poll();
            }
        }
    }

    @Override
    public byte[] diffFrom(UserStream existing) {
        Iterator<UserEvent> iterator = existing.actions.iterator();
        Iterator<UserEvent> myIt = actions.iterator();
        while (iterator.hasNext()) {
            iterator.next();
            myIt.next();
        }

        UserInputPB.UserMessage.Builder output = UserInputPB.UserMessage.newBuilder();

        while (myIt.hasNext()) {
            UserEvent next = myIt.next();
            switch (next.type()) {
                case RESIZE_TYPE -> {
                    UserEvent.Resize resizeEvent = next.as();
                    UserInputPB.Instruction.Builder instructionBuilder = UserInputPB.Instruction.newBuilder();
                    UserInputPB.ResizeMessage.Builder resizeBuilder = UserInputPB.ResizeMessage.newBuilder();
                    resizeBuilder.setWidth(resizeEvent.getWidth());
                    resizeBuilder.setHeight(resizeEvent.getHeight());
                    instructionBuilder.setExtension(UserInputPB.resize, resizeBuilder.build());
                    output.addInstruction(instructionBuilder);
                }
                case USER_BYTE_TYPE -> {
                    UserEvent.UserBytes userBytesEvent = next.as();
                    UserInputPB.Instruction.Builder instructionBuilder = UserInputPB.Instruction.newBuilder();
                    UserInputPB.Keystroke.Builder keystrokeBuilder = UserInputPB.Keystroke.newBuilder();
                    keystrokeBuilder.setKeys(ByteString.copyFrom(userBytesEvent.bytes));
                    instructionBuilder.setExtension(UserInputPB.keystroke, keystrokeBuilder.build());
                    output.addInstruction(instructionBuilder);
                }
            }
        }
        return output.build().toByteArray();
    }

    @Override
    public void pushBack(UserEvent event) {
        actions.offer(event);
    }

    @Override
    public UserStream copy() {
        UserStream userStream = new UserStream();
        userStream.actions.addAll(actions);
        return userStream;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (!(obj instanceof UserStream us)) {
            return false;
        }
        if (this.actions.size() != us.actions.size()) {
            return false;
        }

        Iterator<UserEvent> thisIterator = actions.iterator();
        Iterator<UserEvent> usIterator = us.actions.iterator();
        while (thisIterator.hasNext()) {
            UserEvent thisNext = thisIterator.next();
            UserEvent usNext = usIterator.next();
            if (!thisNext.equals(usNext)) {
                return false;
            }
        }
        return true;
    }

    @Override
    public int actionSize() {
        return actions.size();
    }
}

File: mosh/src/statesync/user_stream.rs
#![allow(dead_code)]
use protobuf::Message;

use super::{UserEvent, UserEventType};
use crate::proto::userinput;

const KEYSTROKE_NUMBER: u32 = 2;
const RESIZE_NUMBER: u32 = 3;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct UserStream {
    actions: Vec<UserEvent>,
}

impl UserStream {
    pub fn new() -> Self {
        UserStream { actions: vec![] }
    }

    pub fn subtract(&mut self, prefix: &UserStream) {
        if self == prefix {
            self.actions.clear();
            return;
        }

        for next in prefix.actions.iter() {
            let mut peek = self.actions.first();
            if let Some(peek) = peek.take() {
                if peek == next {
                    self.actions.remove(0);
                }
            }
        }
    }

    pub fn diff_from(&self, existing: &UserStream) -> Vec<u8> {
        let mut iter = existing.actions.iter();
        let mut my_it = self.actions.iter();
        while let Some(_) = iter.next() {
            my_it.next();
        }

        let mut user_message = userinput::UserMessage::new();

        while let Some(next) = my_it.next() {
            match next.event_type() {
                UserEventType::ResizeType => {
                    let resize_event = next.to_resize();
                    let mut instruction = userinput::Instruction::new();
                    let mut resize = userinput::ResizeMessage::new();
                    resize.set_width(resize_event.width());
                    resize.set_height(resize_event.height());
                    instruction.mut_unknown_fields().add_length_delimited(
                        RESIZE_NUMBER,
                        resize
                            .write_to_bytes()
                            .expect("`ResizeMessage` write to bytes failed."),
                    );
                    user_message.instruction.push(instruction);
                }
                UserEventType::UserByteType => {
                    let user_bytes_event = next.to_user_bytes();
                    let mut instruction = userinput::Instruction::new();
                    let mut keystroke = userinput::Keystroke::new();
                    keystroke.set_keys(user_bytes_event.bytes());
                    instruction.mut_unknown_fields().add_length_delimited(
                        KEYSTROKE_NUMBER,
                        keystroke
                            .write_to_bytes()
                            .expect("`Keystoke` write to bytes failed."),
                    );
                    user_message.instruction.push(instruction);
                }
            }
        }

        user_message
            .write_to_bytes()
            .expect("`UserMessage` write to bytes failed.")
    }

    pub fn push_back(&mut self, event: UserEvent) {
        self.actions.push(event);
    }

    pub fn copy(&self) -> UserStream {
        UserStream {
            actions: self.actions.clone(),
        }
    }

    pub fn action_size(&self) -> usize {
        self.actions.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_protobuf() {
        let mut user_message = userinput::UserMessage::new();
        let mut instruction = userinput::Instruction::new();
        let mut keystroke = userinput::Keystroke::new();
        keystroke.set_keys(vec![0u8; 3]);
        let mut resize = userinput::ResizeMessage::new();
        resize.set_width(100);
        resize.set_height(100);
        instruction.mut_unknown_fields().add_length_delimited(
            KEYSTROKE_NUMBER,
            keystroke
                .write_to_bytes()
                .expect("`Keystoke` write to bytes failed."),
        );
        instruction.mut_unknown_fields().add_length_delimited(
            RESIZE_NUMBER,
            resize
                .write_to_bytes()
                .expect("`ResizeMessage` write to bytes failed."),
        );
        user_message.instruction.push(instruction);

        let bytes = user_message
            .write_to_bytes()
            .expect("`UserMessage` write to bytes failed.");

        let user_message = userinput::UserMessage::parse_from_bytes(&bytes)
            .expect("`UserMessage` parse from bytes failed.");
        user_message.instruction.iter().for_each(|ins| {
            let resize = userinput::exts::resize
                .get(ins)
                .expect("`Instruction` get ResizeMessage failed.");
            assert_eq!(100, resize.width());
            assert_eq!(100, resize.height());

            let keystroke = userinput::exts::keystroke
                .get(ins)
                .expect("`Instruction` get Keystroke failed.");
            assert_eq!(3, keystroke.keys().len());
        });
    }
}
